src/main/java/com/bloxbean/cardano/client/coinselection/impl/DefaultUtxoSelectionStrategyImpl.java

```diff
@@ -1,116 +1,149 @@
+        this.utxoService = utxoService;
+        this.ignoreUtxosWithDatumHash = ignoreUtxosWithDatumHash;
     }
 
     @Override
-    public List<Utxo> selectUtxos(String address, String unit, BigInteger amount, String datumHash, Set<Utxo> excludeUtxos) throws ApiException {
-        if(amount == null)
-            amount = BigInteger.ZERO;
-
-        BigInteger totalUtxoAmount = BigInteger.valueOf(0);
-        List<Utxo> selectedUtxos = new ArrayList<>();
-        boolean canContinue = true;
-        int i = 1;
-
-        while(canContinue) {
-            Result<List<Utxo>> result = utxoService.getUtxos(address, getUtxoFetchSize(),
-                    i++, getUtxoFetchOrder());
-            if(result.code() == 200) {
-                List<Utxo> fetchData = result.getValue();
-
-                List<Utxo> data = filter(fetchData);
-                if(data == null || data.isEmpty()) {
-                    canContinue = false; //Result code 200, but no utxo returned
-                    throw new ApiException(String.format("Unable to get enough Utxos for address : %s, reason: %s", address, result.getResponse()));
-                }
-
-                for(Utxo utxo: data) {
-                    if(excludeUtxos.contains(utxo))
+    public Set<Utxo> select(String sender, List<Amount> outputAmounts, String datumHash, Set<Utxo> utxosToExclude, int maxUtxoSelectionLimit) {
+        if(outputAmounts == null || outputAmounts.isEmpty()){
+            return Collections.emptySet();
+        }
+        try{
+            // loop over utxo's, find matching requested amount
+            Set<Utxo> selectedUtxos = new HashSet<>();
+
+            final Map<String, BigInteger> remaining = new HashMap<>(outputAmounts.stream()
+                    .collect(Collectors.groupingBy(Amount::getUnit,
+                            Collectors.reducing(BigInteger.ZERO,
+                                    Amount::getQuantity,
+                                    BigInteger::add))))
+                    .entrySet().stream()
+                    .filter(entry -> BigInteger.ZERO.compareTo(entry.getValue()) < 0)
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            int page = 0;
+            final int nrOfItems = 100;
+
+            while(!remaining.isEmpty()){
+                var fetchResult = utxoService.getUtxos(sender, nrOfItems, page + 1, OrderEnum.asc);
+
+                var fetched = fetchResult != null && fetchResult.getValue() != null
+                        ? fetchResult.getValue()
+                                     .stream()
+                                     .sorted(sortByMostMatchingAssets(outputAmounts))
+                                     .collect(Collectors.toList())
+                        : Collections.<Utxo>emptyList();
+
+                page += 1;
+                for(Utxo utxo : fetched) {
+                    if(!accept(utxo)){
```

Comment:
What's the use of this condition? It always returns true;

Reply:
:-) Yes, just kept this, might be useful to override a strategy and filter out some UTXOs?